irohad/consensus/yac/impl/yac.cpp

```diff
@@ -136,55 +142,77 @@ namespace iroha {
+              }
 
-          auto processing_state =
-              vote_storage_.getProcessingState(proposal_round);
-
-          auto votes = [](const auto &state) { return state.votes; };
-
-          switch (processing_state) {
-            case ProposalState::kNotSentNotProcessed:
-              vote_storage_.nextProcessingState(proposal_round);
-              log_->info("Propagate state {} to whole network", proposal_round);
-              this->propagateState(visit_in_place(answer, votes));
-              break;
-            case ProposalState::kSentNotProcessed:
-              vote_storage_.nextProcessingState(proposal_round);
-              log_->info("Pass outcome for {} to pipeline", proposal_round);
-              this->closeRound();
-              notifier_.get_subscriber().on_next(answer);
-              break;
-            case ProposalState::kSentProcessed:
-              if (state.size() == 1) {
-                this->findPeer(state.at(0)) | [&](const auto &from) {
-                  log_->info("Propagate state {} directly to {}",
-                             proposal_round,
-                             from->address());
-                  this->propagateStateDirectly(*from,
-                                               visit_in_place(answer, votes));
-                };
+              auto processing_state =
+                  vote_storage_.getProcessingState(proposal_round);
+
+              auto votes = [](const auto &state) { return state.votes; };
+
+              switch (processing_state) {
+                case ProposalState::kNotSentNotProcessed:
+                  vote_storage_.nextProcessingState(proposal_round);
+                  log_->info("Propagate state {} to whole network",
+                             proposal_round);
+                  this->propagateState(visit_in_place(answer, votes));
+                  break;
+                case ProposalState::kSentNotProcessed:
+                  vote_storage_.nextProcessingState(proposal_round);
+                  log_->info("Pass outcome for {} to pipeline", proposal_round);
+                  this->closeRound();
+                  notifier_.get_subscriber().on_next(answer);
+                  break;
+                case ProposalState::kSentProcessed:
+                  tryPropagateBack(state);
+                  break;
               }
-              break;
+            },
+            // sent a state which didn't match with current one
+            [&]() { tryPropagateBack(state); });
+      }
+
+      void Yac::tryPropagateBack(const std::vector<VoteMessage> &state) {
+        // yac back propagation will work  only if another peer is in
+        // propagation stage because if peer sends list of votes this means that
+        // state is already committed
+        if (state.size() != 1) {
+          return;
+        }
+
+        vote_storage_.getLastFinalizedRound() | [&](const auto &last_round) {
+          if (getRound(state) <= last_round) {
```

Comment:
It would be nice to add some logging to both branches of this `if`.

Reply:
What is the point to do that?

Reply:
keysManagerLogger

Reply:
It may be very helpful for synchronization debugging. Much more helpful than other log messages that we currently have.